### PR TITLE
[lab-1] Added pingpong, dump, dump2

### PR DESCRIPTION
**Pull request by Arslanka**

### CODE REVIEW

File: kernel/buddy.c [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+void buddy_bit_inverse(char *array, int index) {
+  char b = array[index / 16];
+  char m = (1 << ((index % 16) / 2));
+  array[index / 16] = (b ^ m);
+}
+
+int buddy_bit_isset(char *array, int index) {
+ char b = array[index / 16];
```
Comment (vityaman) [inline kernel/buddy.c:68]: not clear why `array[index / 16]` here. It looks like `bit_set`, but in `bit_set` logic is understandable, here is not

File: kernel/buddy.c [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+  }
+  printf("\n");
+}
+
+// Print buddy's data structures
+void bd_print() { // doesn't work properly
+  for (int k = 0; k < nsizes; k++) {
+    printf("size %d (blksz %d nblk %d): free list: ", k, BLK_SIZE(k), NBLK(k));
+    lst_print(&bd_sizes[k].free);
+    printf("  alloc:");
```
Comment (vityaman) [inline kernel/buddy.c]: not alloc, but xor_alloc_buddy

File: kernel/file.c [hunk 3]
```diff
@@ -59,27 +55,26 @@ filedup(struct file *f)
 void
 fileclose(struct file *f)
 {
-  struct file ff;
-
   acquire(&ftable.lock);
```
Comment (vityaman) [inline kernel/file.c]: Do we really need to acquire this lock here?

File: kernel/proc.c [hunk 4]
```diff
@@ -89,6 +89,22 @@ myproc(void)
...
 
+// Return the struct proc * by pid, or zero if none.
+static struct proc*
+proc_by_pid(int pid)
+{ 
+  struct proc *caller_proc = myproc();
+  struct proc * p;
+
+  for(p = proc; p < &caller_proc[NPROC]; p++) {
+      if (p->pid == pid && !proc->killed) {
```
Comment (vityaman) [inline kernel/proc.c:100]: race condition is here

File: kernel/sysproc.c [hunk 5]
```diff
@@ -89,3 +89,22 @@ sys_uptime(void)
...
+{
+  return dump();
+}
+
+uint64
+sys_dump2(void)
+{
+  int pid;
+  int register_num;
+  uint64 return_value = 0;
```
Comment (vityaman) [inline kernel/sysproc.c]: why return value is zero initially, while other variables uninitialized?

File: user/pingpong.c [hunk 6]
```diff
@@ -1,4 +1,34 @@
...
+    }
+
+    int pid = fork();
+
+    if (pid < 0) {
+        printf("fork failed\n");
+        exit(1);
+    } else if (pid == 0) { // child 
+        char buf[8];
+        read(fd[0], buf, sizeof buf);
```
Comment (vityaman) [inline user/pingpong.c:18]: It is better to handle errors, not ignoring them

**Review by vityaman — CHANGES_REQUESTED** [review]
Not so bad, but some small changes are needed

**Comment by vityaman** [comment]
Ping

**Comment by Arslanka** [comment]
> Ping

Pong

File: kernel/proc.c (hunk 4 above)
Comment (vityaman) [reply kernel/proc.c:100]: wtf?! It should not work!

Comment (vityaman) [reply kernel/proc.c:100]: Here, caller_proc -- myproc, not an array,
proc is an array. Also proc->killed is an mistake p->killed is correct

File: kernel/file.c (hunk 3 above)
Comment (vityaman) [reply kernel/file.c]: It is better to lock only file, not all files in the system

**Pull request merged**